### PR TITLE
Dev real ping

### PR DESCRIPTION
**Pull request by DuckVador**
- [x] Test on Windows 

- [x] Test on macOS

### CODE REVIEW

**Comment by DuckSoft** [comment]
Review in Progress on Linux.

**Comment by U-v-U** [comment]
Rebased to the latest dev / version-v2 / master branch.

**Comment by DuckSoft** [comment]
Tested on Linux. The reading isn't very stable.
Also, I wonder if we should split latency checking method into 3 ones.
I think there should be one option like "Use Real Ping On Connected Node".

What do you think?

**Comment by DuckVador** [comment]
> Tested on Linux. The reading isn't very stable.
> Also, I wonder if we should split latency checking method into 3 ones.
> I think there should be one option like "Use Real Ping On Connected Node".
> 
> What do you think?
I think it is feasible because the three do not conflict. Are there too many right-click menu options?

File: .github/workflows/nsis.yml [hunk 1]
```diff
@@ -85,11 +85,18 @@ jobs:
         shell: bash
         run: |
          curl -o ./libs/Qv2ray-deps-grpc-${{ matrix.arch }}-windows.7z -L https://github.com/Qv2ray/Qv2ray-deps/releases/download/release/Qv2ray-deps-grpc-${{ matrix.arch }}-windows.7z
+         curl -o ./libs/Qv2ray-deps-curl-${{ matrix.arch }}-windows.7z -L https://github.com/DuckVador/Qv2ray-deps/releases/download/release/Qv2ray-deps-curl-${{ matrix.arch }}-windows.7z
```
Comment (U-v-U) [inline .github/workflows/nsis.yml:88]: Can this be moved into Qv2ray org?

**Comment by U-v-U** [comment]
Just one change requested, @DuckVador

**Comment by U-v-U** [comment]
Waiting for https://github.com/Qv2ray/Qv2ray-deps/actions/runs/202169046

**Comment by DuckSoft** [comment]
This will introduce an extra dependency. Do not merge until confirmed by downstream packagers.

- [ ] AUR: @DuckSoft 
- [ ] Arch Linux CN: @DuckSoft 
- [ ] Debian: @ymshenyu 
- [ ] openSUSE / Fedora: @zzndb 
- [ ] Gentoo: @12101111

**Comment by U-v-U** [comment]
How about macOS?
I didn't see a dylib is installed into macOS bundle, is libcurl statically linked?
https://github.com/Qv2ray/Qv2ray/runs/965726214?check_suite_focus=true#step:17:550

**Comment by DuckVador** [comment]
> How about macOS?
> I didn't see a dylib is installed into macOS bundle, is libcurl statically linked?
> https://github.com/Qv2ray/Qv2ray/runs/965726214?check_suite_focus=true#step:17:550

curl is installed by default on macOS.

**Comment by DuckSoft** [comment]
@DuckVador Will there be ABI break between different bugMac versions?

**Comment by DuckVador** [comment]
> @DuckVador Will there be ABI break between different bugMac versions?



[curl abi history](url)

> The previous major library soname number bumps (breaking backwards compatibility) have happened the following times:
0 - libcurl 7.1, August 2000
1 - libcurl 7.5 December 2000
2 - libcurl 7.7 March 2001
3 - libcurl 7.12.0 June 2004
4 - libcurl 7.16.0 October 2006

macOS 10.14 curl version is 7.54. Seems no breaking changes.

**Comment by U-v-U** [comment]
LGTM

**Comment by U-v-U** [comment]
Waiting for @zzndb and @12101111

File: src/components/latency/RealPing.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        auto request_name = GlobalConfig.networkConfig.latencyRealPingTestURL.toStdString();
+        data.totalCount = 0;
+        data.failedCount = 0;
+        data.worst = 0;
+        data.avg = 0;
+        auto local_proxy_address =
+            (!GlobalConfig.inboundConfig.useHTTP ?
+                 "socks5://" + GlobalConfig.inboundConfig.listenip + ":" + QSTRN(GlobalConfig.inboundConfig.socksSettings.port) :
+                 "http://" + GlobalConfig.inboundConfig.listenip + ":" + QSTRN(GlobalConfig.inboundConfig.httpSettings.port))
+                .toStdString();
```
Comment (DuckSoft) [inline src/components/latency/RealPing.cpp:164]: WARNING: You may need to process IPv6 addresses.

**Pull request merged**